dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java

```diff
@@ -175,26 +200,64 @@ private void handleDataValues(
 
           if (isNewDataValue(dbDataValue, dataValue)) {
             eventChangeLogService.addDataValueChangeLog(
-                event, dataElement, dataValue.getValue(), null, CREATE, user.getUsername());
+                event, dataElement, null, dataValue.getValue(), CREATE, user.getUsername());
             saveDataValue(dataValue, event, dataElement, user, entityManager, preheat);
           } else if (isUpdate(dbDataValue, dataValue)) {
             eventChangeLogService.addDataValueChangeLog(
                 event,
                 dataElement,
-                dataValue.getValue(),
                 dbDataValue.getValue(),
+                dataValue.getValue(),
                 UPDATE,
                 user.getUsername());
             updateDataValue(
                 dbDataValue, dataValue, event, dataElement, user, entityManager, preheat);
           } else if (isDeletion(dbDataValue, dataValue)) {
             eventChangeLogService.addDataValueChangeLog(
-                event, dataElement, null, dbDataValue.getValue(), DELETE, user.getUsername());
+                event, dataElement, dbDataValue.getValue(), null, DELETE, user.getUsername());
             deleteDataValue(dbDataValue, event, dataElement, entityManager, preheat);
           }
         });
   }
 
+  private void logEventPropertyChanges(Event originalEvent, Event payloadEvent, UserDetails user) {
+    logIfChanged(
+        "occurredDate",
+        Event::getOccurredDate,
+        this::formatDate,
+        originalEvent,
+        payloadEvent,
+        user);
+    logIfChanged(
+        "scheduledDate",
+        Event::getScheduledDate,
+        this::formatDate,
+        originalEvent,
+        payloadEvent,
+        user);
+    logIfChanged(
+        "geometry", Event::getGeometry, this::formatGeometry, originalEvent, payloadEvent, user);
+  }
+
+  private <T> void logIfChanged(
+      String propertyName,
+      Function<Event, T> valueExtractor,
+      Function<T, String> formatter,
+      Event originalEvent,
+      Event payloadEvent,
+      UserDetails user) {
+
+    String originalValue = formatter.apply(valueExtractor.apply(originalEvent));
+    String newValue = formatter.apply(valueExtractor.apply(payloadEvent));
+
+    if (!Objects.equals(originalValue, newValue)) {
```

Comment:
The changelog service API looks too complex to me. I think we should push down as much as we can into the service.

Can the service not
* discard any calls that try to log unchanged values?
* figure out the changelog type?
* figure out the username?

if it can it should (unless I am missing something).

```java
eventChangeLogService.addEventPropertyChangeLog(event, property, previousValue, newValue);
```

(same for data values)